File: src/main/kotlin/ktln/lib/SegmentTreeR.kt
package ktln.lib

/**
 * セグメント木
 *
 * @property arr 対象のリスト
 * @property idElm 単位元(演算メソッドで結果がもう一方の値と同じとなる要素。足し算なら0、掛け算なら1。)
 * @property operator 演算メソッド(合計, 最大値, 最小値などの計算)
 */
class SegmentTreeR(
  private val arr: List<Long>,
  private val idElm: Long,
  private val operator: (Long, Long) -> Long
) {
  public val leafSize: Int
  private val tree: MutableList<Long>

  init {
    var i = 1
    while (i < arr.size) i *= 2
    leafSize = i
    tree = MutableList(leafSize * 2) { idElm }
    for (j in arr.indices) tree[leafSize + j] = arr[j]
    for (k in leafSize - 1 downTo 1) tree[k] = operator(tree[2 * k], tree[2 * k + 1])
  }

  /**
   * 更新
   *
   * @param pos arrのindex + 1
   * @param x 更新する値
   */
  fun update(pos: Int, x: Long) {
    var idx = pos + leafSize - 1
    tree[idx] = x
    while (idx > 1) {
      idx /= 2
      tree[idx] = operator(tree[2 * idx], tree[2 * idx + 1])
    }
  }

  /**
   * 値の取得
   *
   * @param l 求めたい半開区間の左
   * @param r 求めたい半開区間の右（その値は含まない）
   * @param a 現在の半開区間の左
   * @param b 現在の半開区間の右（その値は含まない）
   * @param idx 現在のtreeのインデックス
   * @return
   */
  fun query(l: Int, r: Int, a: Int = 1, b: Int = leafSize + 1, idx: Int = 1): Long  {
    if (r <= a || b <= l) return idElm
    if (l <= a && b <= r) return tree[idx]
    val mid = (a + b) / 2
    return operator(query(l, r, a, mid, idx * 2), query(l, r, mid, b, idx * 2 + 1))
  }
}

File: ruby/lib/segment_tree_r.rb
# セグメント木

class SegmentTreeR
  # leaf_size: 葉の数
  # ope: 演算メソッド(合計, 最大値, 最小値などの計算)
  # id_elm: 単位元(演算メソッドで結果がもう一方の値と同じとなる要素。足し算なら0、掛け算なら1。)
  # tree: 全体の配列(インデックス0は使用しない)
  attr_accessor :leaf_size, :ope, :id_elm, :tree

  def initialize(arr, id_elm, &method)
    n = arr.size
    @ope = method
    @id_elm = id_elm
    @leaf_size = 1 << (n - 1).bit_length
    @tree = [id_elm] * 2 * leaf_size
    arr.each_with_index {|a, i| tree[leaf_size + i] = a }
    (leaf_size - 1).downto(1) { update(_1) }
  end

  # 更新
  # pos arrのindex
  # value 更新する値
  def set(pos, value)
    idx = leaf_size + pos
    tree[idx] = value
    while idx > 1
      update(idx >>= 1)
    end
  end

  # 値の取得
  # [l, r)　求めたい半開区間
  # [a, b) 現在の半開区間
  # u 現在のtreeのインデックス
  def prod(l, r, a = 0, b = leaf_size, idx = 1)
    return id_elm if r <= a || b <= l
    return tree[idx] if l <= a && b <= r
    mid = (a + b) / 2
    res_l = prod(l, r, a, mid, idx * 2)
    res_r = prod(l, r, mid, b, idx * 2 + 1)
    ope.call(res_l, res_r)
  end

  private
  def update(idx)
    tree[idx] = ope.call(tree[2 * idx], tree[2 * idx + 1])
  end
end
